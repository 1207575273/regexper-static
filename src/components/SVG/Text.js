import React from 'react';
import PropTypes from 'prop-types';

import style from './style';

import reflowable from './reflowable';

@reflowable
class Text extends React.PureComponent {
  reflow() {
    const box = this.text.getBBox();

    this.setBBox({
      width: box.width,
      height: box.height
    });

    return this.setStateAsync({
      transform: `translate(${-box.x} ${-box.y})`
    });
  }

  textRef = text => this.text = text

  renderContent() {
    const { children, quoted } = this.props;
    if (!quoted) {
      return children;
    }

    return <React.Fragment>
      <tspan style={ style.textQuote }>&ldquo;</tspan>
      <tspan>{ children }</tspan>
      <tspan style={ style.textQuote }>&rdquo;</tspan>
    </React.Fragment>;
  }

  render() {
    const { theme } = this.props;
    const { transform } = this.state || {};

    const textProps = {
      style: { ...style.text, ...style[theme] },
      transform,
      ref: this.textRef
    };

    return <text { ...textProps }>
      { this.renderContent() }
    </text>;
  }
}

Text.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
  quoted: PropTypes.bool,
  theme: PropTypes.string
};

export default Text;
