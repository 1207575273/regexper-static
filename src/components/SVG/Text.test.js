import React from 'react';
import { mount } from 'enzyme';

import Text from './Text';

const originalGetBBox = window.Element.prototype.getBBox;

describe('Text', () => {
  beforeEach(() => {
    window.Element.prototype.getBBox = function() {
      return { x: 10, y: 10 };
    };
  });

  afterEach(() => {
    window.Element.prototype.getBBox = originalGetBBox;
  });

  test('rendering', async () => {
    const component = mount(
      <Text>Test content</Text>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });

  test('rendering with quotes', async () => {
    const component = mount(
      <Text quoted>Test content</Text>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });
});
