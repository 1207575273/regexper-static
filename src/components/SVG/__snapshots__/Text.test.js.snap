// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text rendering 1`] = `
<Text>
  <text
    style={
      Object {
        "fontFamily": "Arial",
        "fontSize": "16px",
      }
    }
    transform="translate(-10 -10)"
  >
    Test content
  </text>
</Text>
`;

exports[`Text rendering with quotes 1`] = `
<Text
  quoted={true}
>
  <text
    style={
      Object {
        "fontFamily": "Arial",
        "fontSize": "16px",
      }
    }
    transform="translate(-10 -10)"
  >
    <tspan
      style={
        Object {
          "fill": "#908c83",
        }
      }
    >
      “
    </tspan>
    <tspan>
      Test content
    </tspan>
    <tspan
      style={
        Object {
          "fill": "#908c83",
        }
      }
    >
      ”
    </tspan>
  </text>
</Text>
`;
